fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/postgresql/PostgreSqlCodeSystemDAO.java

```diff
@@ -21,84 +23,95 @@
+     */
     @Override
-    public int readOrAddCodeSystem(String codeSystem) throws FHIRPersistenceDataAccessException   {
-        // As the system is concurrent, we have to handle cases where another thread
-        // might create the entry after we selected and found nothing
-        Integer result = getCodeSystemId(codeSystem);
-
-        // Create the resource if we don't have it already (set by the continue handler)
-        if (result == null) {
-            try {
-                result = fhirRefSequenceDAO.nextValue();
-
-                String INS = "INSERT INTO code_systems (code_system_id, code_system_name) VALUES (?, ?)";
-                try (PreparedStatement stmt = getConnection().prepareStatement(INS)) {
-                    // bind parameters
-                    stmt.setInt(1, result);
-                    stmt.setString(2, codeSystem);
-                    stmt.executeUpdate();
+    public int readOrAddCodeSystem(String systemName) throws FHIRPersistenceDataAccessException   {
+        final String METHODNAME = "readOrAddCodeSystem";
+        log.entering(CLASSNAME, METHODNAME);
+
+        int systemId;
+        String currentSchema;
+        String stmtString;
+        String errMsg = "Failure storing code system id: name=" + systemName;
+        long dbCallStartTime;
+        double dbCallDuration;
+
+        try {
+            currentSchema = getConnection().getSchema().trim();
+            stmtString = String.format(SQL_CALL_ADD_CODE_SYSTEM_ID, currentSchema);
+            try (CallableStatement stmt = getConnection().prepareCall(stmtString)) {
+                stmt.setString(1, systemName);
+                stmt.registerOutParameter(2, Types.INTEGER);
+                dbCallStartTime = System.nanoTime();
+                stmt.execute();
+                dbCallDuration = (System.nanoTime()-dbCallStartTime)/1e6;
+                if (log.isLoggable(Level.FINE)) {
+                        log.fine("DB read code system id complete. executionTime=" + dbCallDuration + "ms");
                 }
+                systemId = stmt.getInt(2);
             }
-            catch (SQLException e) {
-                if ("23505".equals(e.getSQLState())) {
-                    // another thread snuck in and created the record, so we need to fetch the correct id
-                    result = getCodeSystemId(codeSystem);
-
-                    if (result == null) {
-                        // This would be truly weird, but we protect against it anyway
-                        throw new IllegalStateException("No code system returned after duplicate found!");
-                    }
-                }
-                else {
-                    throw new FHIRPersistenceDataAccessException("codeSystem=" + codeSystem, e);
-                }
-            }
-
         }
-
-        // There's no way result can be null here, so we're OK returning an int
-        return result;
+        catch (Throwable e) {
```

Comment:
merge 73, 74 to oneline

Reply:
followed Robin's code style, let me change. :)